File: desktop-workbench/renderer/test/timeline-timestamp.test.mjs
/**
 * Timestamp formatting for timeline entries.
 *
 * A transcript can span hours, and without a visible time a reader cannot tell
 * when a reply or an action began. These cases pin the exact rendered shape,
 * because the value is read as data: two timestamps must be comparable at a
 * glance regardless of the reader's locale.
 */

import assert from "node:assert/strict"
import test from "node:test"

import {
  formatTimelineTimestamp,
  localDateTime,
} from "../src/components/session/timeline-timestamp.ts"

test("formats an instant as a bracketed local date and time", () => {
  const date = new Date(2026, 8, 16, 14, 5, 9) // 2026-09-16 14:05:09 local
  const formatted = localDateTime(date)
  assert.equal(formatted, "2026-09-16 14:05:09")
})

test("pads every field to a fixed width", () => {
  const date = new Date(2026, 0, 2, 3, 4, 5) // 2026-01-02 03:04:05 local
  assert.equal(localDateTime(date), "2026-01-02 03:04:05")
})

test("wraps the timestamp in square brackets", () => {
  const iso = new Date(2026, 8, 16, 14, 5, 9).toISOString()
  const formatted = formatTimelineTimestamp(iso)
  assert.ok(formatted, "a valid instant must format")
  assert.match(formatted, /^\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\]$/)
})

test("renders nothing for a missing timestamp", () => {
  assert.equal(formatTimelineTimestamp(null), null)
  assert.equal(formatTimelineTimestamp(undefined), null)
  assert.equal(formatTimelineTimestamp(""), null)
})

test("renders nothing rather than an invalid date", () => {
  // A placeholder must not become the literal text "Invalid Date".
  assert.equal(formatTimelineTimestamp("not-a-time"), null)
  assert.equal(formatTimelineTimestamp("2026-13-45T99:99:99Z"), null)
})

test("uses the same format for every timestamp so they line up", () => {
  const first = formatTimelineTimestamp(new Date(2026, 8, 16, 9, 0, 0).toISOString())
  const second = formatTimelineTimestamp(new Date(2026, 8, 16, 17, 30, 0).toISOString())
  assert.ok(first && second)
  assert.equal(first.length, second.length)
})

File: android/app/src/test/java/com/agentsanywhere/app/feature/sessiondetail/TimelineTimestampTest.kt
package com.agentsanywhere.app.feature.sessiondetail

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.ZoneId

/**
 * Timestamp formatting for timeline rows.
 *
 * A transcript can span hours, and without a visible time a reader cannot tell
 * when a reply or an action began. These cases pin the exact rendered shape,
 * because the value is read as data: two timestamps must be comparable at a
 * glance regardless of the device language.
 */
class TimelineTimestampTest {
    private val utc = ZoneId.of("UTC")

    @Test
    fun `formats an instant as a bracketed month, day and time`() {
        assertEquals(
            "[09-16 14:05]",
            formatTimelineTimestamp("2026-09-16T14:05:09Z", utc),
        )
    }

    @Test
    fun `omits seconds and the year`() {
        val formatted = formatTimelineTimestamp("2026-09-16T14:05:09Z", utc)
        assertTrue("the year must not appear", formatted?.contains("2026") != true)
        assertEquals("only hours and minutes", 2, formatted!!.removePrefix("[").removeSuffix("]").split(":").size)
    }

    @Test
    fun `pads every field to a fixed width`() {
        assertEquals(
            "[01-02 03:04]",
            formatTimelineTimestamp("2026-01-02T03:04:05Z", utc),
        )
    }

    @Test
    fun `renders the instant in the supplied zone`() {
        // 14:05 UTC is 22:05 in Beijing, which is what a reader there expects.
        assertEquals(
            "[09-16 22:05]",
            formatTimelineTimestamp("2026-09-16T14:05:09Z", ZoneId.of("Asia/Shanghai")),
        )
    }

    @Test
    fun `renders nothing for a missing timestamp`() {
        assertNull(formatTimelineTimestamp(null, utc))
        assertNull(formatTimelineTimestamp("", utc))
        assertNull(formatTimelineTimestamp("   ", utc))
    }

    @Test
    fun `renders nothing rather than an unparseable value`() {
        assertNull(formatTimelineTimestamp("not-a-time", utc))
        assertNull(formatTimelineTimestamp("2026-13-45T99:99:99Z", utc))
    }

    @Test
    fun `uses the same width for every timestamp so they line up`() {
        val first = formatTimelineTimestamp("2026-09-16T09:00:00Z", utc)
        val second = formatTimelineTimestamp("2026-09-16T17:30:00Z", utc)
        assertTrue(first != null && second != null)
        assertEquals(first!!.length, second!!.length)
        // Short enough to sit above a row without dominating it.
        assertTrue("stamp is too long: $first", first.length <= 16)
    }
}
